File: oop1/Character.cpp
#include "Character.h"

#include <iostream>

using std::cout;
using std::endl;

Character::Character(const std::string& name, int level)
    : name(name),
      level(level),
      health(100 * level),
      maxHealth(100 * level),
      baseDamage(20 * (level * 2)),
      regenAmount(5),
      regenIntervalSec(5.0f),
      regenAccumulator(0.0f) {}

void Character::useSpecial() {
    cout << name << " не имеет специального умения." << endl;
}

void Character::protectAlly(Character& ally, int incomingDamage) {
    cout << name << " не может защитить союзника. Урон получает "
         << ally.getName() << "." << endl;
    ally.takeDamage(incomingDamage);
}

void Character::takeDamage(int damage) {
    health -= damage;
    if (health < 0) health = 0;
    cout << name << " получает " << damage << " урона. Осталось: "
         << health << "/" << maxHealth << endl;
}

void Character::heal(int amount) {
    health += amount;
    if (health > maxHealth) health = maxHealth;
    cout << name << " восстанавливает " << amount << " здоровья." << endl;
}

void Character::update(float deltaSeconds) {
    if (health <= 0) return;
    onUpdate(deltaSeconds);
    regenAccumulator += deltaSeconds;
    while (regenAccumulator >= regenIntervalSec) {
        regenAccumulator -= regenIntervalSec;
        heal(regenAmount);
    }
}

void Character::printInfo() const {
    cout << "Персонаж: " << name << ", уровень: " << level
         << ", здоровье: " << health << "/" << maxHealth
         << ", базовый урон: " << baseDamage << endl;
}

std::string Character::getName() const { return name; }
int Character::getHealth() const { return health; }
int Character::getLevel() const { return level; }
int Character::getMaxHealth() const { return maxHealth; }
int Character::getBaseDamage() const { return baseDamage; }

void Character::setHealth(int h) {
    if (h >= 0 && h <= maxHealth) {
        health = h;
    }
}

void Character::setBaseDamage(int dmg) {
    if (dmg > 0) {
        baseDamage = dmg;
    }
}

void Character::onUpdate(float) {}

File: oop3/src/main/java/com/game/entity/Character.java
package com.game.entity;

import jakarta.persistence.Column;
import jakarta.persistence.DiscriminatorColumn;
import jakarta.persistence.DiscriminatorType;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Inheritance;
import jakarta.persistence.InheritanceType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;

import lombok.Getter;
import lombok.Setter;

@Entity
@Table(name = "characters")
@Inheritance(strategy = InheritanceType.JOINED)
@DiscriminatorColumn(name = "character_type", discriminatorType = DiscriminatorType.STRING)
@Getter
@Setter
public abstract class Character {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private Integer level;

    @Column(nullable = false)
    private Integer health;

    @Column(name = "max_health", nullable = false)
    private Integer maxHealth;

    @Column(name = "base_damage", nullable = false)
    private Integer baseDamage;

    @Column(name = "regen_amount", nullable = false)
    private Integer regenAmount;

    @Column(name = "regen_interval_sec", nullable = false)
    private Float regenIntervalSec;

    @ManyToOne
    @JoinColumn(name = "guild_id")
    private Guild guild;

    protected Character() {
    }

    protected Character(Long id, String name, Integer level, Integer health, Integer maxHealth,
      Integer baseDamage, Integer regenAmount, Float regenIntervalSec) {
        this.id = id;
        this.name = name;
        this.level = level;
        this.health = health;
        this.maxHealth = maxHealth;
        this.baseDamage = baseDamage;
        this.regenAmount = regenAmount;
        this.regenIntervalSec = regenIntervalSec;
    }

    public void setHealth(Integer health) {
        this.health = Math.max(0, Math.min(health, maxHealth == null ? health : maxHealth));
    }

    public void setBaseDamage(Integer baseDamage) {
        if (baseDamage != null && baseDamage > 0) {
            this.baseDamage = baseDamage;
        }
    }

    public void useSpecial() {
        System.out.println(name + " не имеет специального умения.");
    }

    public void protectAlly(Character ally, int incomingDamage) {
        System.out.println(name + " не может защитить союзника. Урон получает " + ally.getName() + ".");
        ally.takeDamage(incomingDamage);
    }

    public void takeDamage(int damage) {
        health = Math.max(0, health - damage);
    }

    public void heal(int amount) {
        health = Math.min(maxHealth, health + amount);
    }

    protected void onUpdate(float deltaSeconds) {
    }

    public void update(float deltaSeconds) {
        if (health <= 0) return;
        onUpdate(deltaSeconds);
    }

    public String getCharacterClass() {
        return getClass().getSimpleName();
    }

    @Override
    public String toString() {
        return "%s{id=%d, name='%s', level=%d}".formatted(getCharacterClass(), id, name, level);
    }
}
